job-runner/chain/chain.go

```diff
@@ -18,28 +18,40 @@ type chain struct {
 	jcMux    *sync.RWMutex
 	jobChain *proto.JobChain
 
-	runningMux *sync.RWMutex
-	running    map[string]proto.JobStatus // keyed on job id
-	n          uint                       // N jobs ran
+	runningMux         *sync.RWMutex
+	running            map[string]proto.JobStatus // keyed on job id
+	n                  uint                       // N jobs ran
+	sequenceRetryCount map[string]uint            // Number of retries attempted so far
 }
 
 // NewChain takes a JobChain proto (from the RM) and turns it into a Chain that
 // the JR can use.
 func NewChain(jc *proto.JobChain) *chain {
-	// Set the state of all jobs in the chain to "Pending".
+	sequenceRetryCount := make(map[string]uint)
 	for jobName, job := range jc.Jobs {
+		// Set the state of all jobs in the chain to "Pending".
 		job.State = proto.STATE_PENDING
+
+		// Initialize job.Data
 		job.Data = map[string]interface{}{}
+
+		if job.Sequence.Retry > 0 && jobName == job.Sequence.StartJobId {
```

Comment:
In order to support nested sequence retries, we need to change the way we store info about the Sequence retries. 

Since every job is part of a sequence, we can change to encode parent sequence start id and on the start node of ever sequence, encode the actual retry.

This is different from above because the start node will no longer necessarily have the same jobeName job.Sequence.StartJobId